File: app/src/main/java/com/example/ui/theme/ZadExtendedColors.kt
package com.example.ui.theme

import androidx.compose.runtime.Immutable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color

/**
 * Colours the app needs that Material's ColorScheme has no slot for.
 *
 * `MaterialTheme.colorScheme` covers primary/surface/error and friends, but nothing in
 * it means "success", "tertiary text", or "the middle stop of the canvas gradient".
 * Those lived as top-level vals, which are fixed at class init and therefore invisible
 * to a theme switch — so after the dark scheme landed they stayed light while everything
 * around them went dark. This carries them through the composition instead, so a single
 * value swapped in AppTheme moves all of them at once.
 *
 * `staticCompositionLocalOf` rather than `compositionLocalOf`: the value only changes
 * when the theme itself changes, and that should invalidate the whole subtree anyway.
 */
@Immutable
data class ZadExtendedColors(
    val success: Color,
    val info: Color,
    val textTertiary: Color,
    val primaryDark: Color,
    val primaryLight: Color,
    val secondaryDark: Color,
    val secondaryLight: Color,
    val coral: Color,
    val coralLight: Color,
    val lilac: Color,
    val canvasTop: Color,
    val canvasMid: Color,
    val canvasBottom: Color,
    val surfaceContainerLow: Color,
    val surfaceContainer: Color,
    val surfaceContainerHigh: Color,
    /** Decorative radial washes behind the auth/onboarding canvas. */
    val authWashWarm: Color,
    val authWashCool: Color,
    /**
     * لوحة الأيجنت — **داكنة في الوضعين عن قصد**، مش بتتقلب مع الثيم.
     *
     * الكارت كله نصه `Color.White` ثابت (المخطط بيحدده كلوحة ليلية:
     * تدرّج 135° من #052E16 لـ#0A382C). قبل كده كانت بداية التدرّج
     * `primaryContainer` — رمز بيتقلب — فبقت #E8F0EC في اللايت مود، يعني
     * **أبيض على شبه أبيض بتباين 1.16:1**. الطرف التاني كان ثابت داكن، فالعطل
     * كان في النص التاني من الكارت بس والأول غير مقروء.
     *
     * الرمزين دول بيخلّوا القيمتين في الثيم (فيتغيّروا من مكان واحد) من غير ما
     * يتقلبوا — لأن التقليب هو نفسه اللي كسرها.
     */
    val agentPanelStart: Color,
    val agentPanelEnd: Color,
    /**
     * ألوان الحالة **فوق لوحة الأيجنت** (أيقونات تنبيه، شرائح اقتراحات، رقم «قرب ينتهي»).
     *
     * نفس منطق `agentPanelStart`: اللوحة داكنة في الوضعين، فالرموز دي ماينفعش تتقلب مع
     * الثيم. `successColor` العادي في اللايت (#238652) بيدّي **2.85:1** فوق اللوحة — مش
     * مقروء. فالقيم هنا هي درجات العلامة المصمَّمة للأرضيات الداكنة (success/info بتوع
     * الدارك + الأوكر الفاتح)، ثابتة في الوضعين. بتحلّ محل هيكسات ماتيريال
     * `#4CAF50`/`#FF9800`/`#2196F3` اللي كانت مش من العلامة، و`#2196F3` كان **4.16:1** تحت AA.
     * القياس على أسوأ stop في الجراديانت: warning 6.28، info 5.89، success 5.66.
     */
    val agentPanelSuccess: Color,
    val agentPanelWarning: Color,
    val agentPanelInfo: Color,
    /**
     * هوية زاد فوق اللوحة (نجمة الـsparkle + كلمة «Zad Agent»). كانت `primary`، ودي في
     * اللايت #1B4332 = **1.17:1** فوق اللوحة — مش باينة خالص (اتكشفت من لقطة Roborazzi
     * مش من القراية). المنت ده هو primaryLight بتاع الدارك، والبروتوتايب نفسه كاتبه «mint»: 7.97:1.
     */
    val agentPanelBrand: Color,
    /**
     * بانر تنبيه الذكاء في الرئيسية (`AiAlertBanner`): حاوية + نص.
     *
     * كان `Color(0xFFFDECEA)` ثابت مع نص `dangerColor`، وتعليقه بيقول إن الحاوية
     * المصمتة بتعدّي AA. القياس بيقول غير كده: لايت **3.44:1** (الوصف 12sp عتبته 4.5)،
     * ودارك أسوأ — الكارت بيفضل وردي فاتح والنص بيبقى `#EA764B` = **2.55:1**. فالحاوية
     * والنص بقوا زوج واحد بيتقلب مع بعض: لايت 4.79:1، دارك 5.22:1.
     */
    val alertBannerContainer: Color,
    val onAlertBanner: Color,
    /**
     * هوية الذكاء — البنفسجي الرسمي (قرار المستخدم 2026-09-13، المرحلة ٣).
     *
     * شاشة العقل كانت فيها ٢٣ هيكس بنفسجي من غير أي تعريف: `#9333EA` لوحده ١٢ مرة، وصناديق
     * سرد الذكاء لافندر ثابت `#F5F3FF` بيفضل فاتح في الدارك مود. القرار كان بين إن البنفسجي
     * يتشال لصالح الزمردي أو يتعلن هوية للذكاء — واتعلن. فاللايت بيفضل **نفس القيم بالظبط**
     * (مفيش تغيير بصري)، والدارك بياخد درجات مقروءة على الأرضية الداكنة. ولو القرار اتغيّر
     * بعدين، التحويل للزمردي بيبقى هنا بس.
     *
     * - `aiAccent`: نص/أيقونة/حاوية زرار. لايت #9333EA (5.38 على السطح، 4.51 على شريحة ١٢٪)،
     *   دارك #C084FC (6.45 / 5.37).
     * - `aiAccentEnd`: الطرف التاني من جراديانت الهوية، ثابت.
     * - `onAiAccent`: فوق `aiAccent`. أبيض في اللايت (5.38)، بس الأبيض على بنفسجي الدارك
     *   **2.64:1** — فالدارك نص غامق (6.87، و4.1 على الطرف التاني من الجراديانت).
     * - `aiAccentSoft`: نبضة العقدة في كانفس الشبكة — زخرفة، ثابتة.
     * - `aiNarrativeContainer`/`onAiNarrative`: صندوق سرد الذكاء. لايت 9.99:1، دارك 11.09:1
     *   (الحاوية بقت بنفسجي غامق بدل ما تفضل لافندر فاتح وسط شاشة داكنة).
     */
    val aiAccent: Color,
    val aiAccentEnd: Color,
    val onAiAccent: Color,
    val aiAccentSoft: Color,
    val aiNarrativeContainer: Color,
    val onAiNarrative: Color,
    /**
     * درجة «كويس» في مقياس صحة البيت (بين success وsecondary). كان ليموني Tailwind
     * `#84CC16` = **1.98:1** على الأبيض، والرقم نفسه (22sp Black، نص كبير عتبته 3:1)
     * بيتلوّن بيه. لايت #4D7C0F = 4.99، دارك #A3E635.
     */
    val scoreGood: Color,
    /**
     * نص شريحة الحالة فوق حاوية بنفس درجتها بشفافية ١٢٪.
     *
     * `coral` نفسه مابيعديش هنا: نص بنفس درجة لون الحاوية بيدي **٣٫٣٨:١** في
     * الوضع الفاتح، والشريحة نصها 11sp bold يعني مش «نص كبير» فعتبتها 4.5:1 مش 3:1.
     * القياس على الشريحة الفعلية (coral@12% فوق السطح): لايت 5.40:1، دارك 4.94:1.
     * الرمز ده أغمق من `coral` في اللايت بس، والدارك بيفضل زي ما هو.
     */
    val onStatusPill: Color,
    /**
     * Categorical chart palette — colours whose only job is to tell series apart.
     *
     * Deliberately NOT the semantic tokens. The third slice of a spending breakdown is
     * not "a warning" and the fourth is not "an error"; reusing `warningColor` /
     * `dangerColor` there would attach a meaning the data does not have, and would also
     * make two series collide the moment a semantic token is retuned.
     *
     * Ordered by how distinguishable adjacent entries are, because the consumer takes
     * them in order (`palette[index % size]`) — so neighbouring categories must not look
     * alike.
     */
    val chartCategorical: List<Color>,
    /** Body text on a card that is not the primary label — dimmer than onSurface. */
    val textOnCardSecondary: Color,
)

/** Exactly the values these tokens held before the extended-colour mechanism existed. */
val ZadExtendedColorsLight = ZadExtendedColors(
    success = Color(0xFF238652),
    info = Color(0xFF2B6CB0),
    // كان 9EA197 = 2.49:1 على الخلفية — نص مايتقريش (WCAG AA عايز 4.5). 6E7065 = 4.78:1،
    // ولسه أفتح من onSurfaceVariant فالتلات أوزان محافظين على ترتيبهم (WcagContrastTest).
    textTertiary = Color(0xFF6E7065),
    primaryDark = ZadForestEmeraldDark,
    primaryLight = ZadForestEmeraldLight,
    secondaryDark = ZadMustardDark,
    secondaryLight = ZadMustardLight,
    coral = ZadTerracottaRust,
    coralLight = ZadTerracottaContainer,
    lilac = Color(0xFF7C6F93),
    canvasTop = Color(0xFFF8F9FA),
    canvasMid = Color(0xFFF4F6F2),
    canvasBottom = Color(0xFFEDEFE9),
    surfaceContainerLow = Color(0xFFFBFBFA),
    surfaceContainer = ZadIosBackground,
    surfaceContainerHigh = ZadIosSurfaceVariant,
    authWashWarm = Color(0xFFFCD3C7),
    authWashCool = Color(0xFFBFE3D1),
    agentPanelStart = Color(0xFF052E16),
    agentPanelEnd = Color(0xFF0A382C),
    agentPanelSuccess = Color(0xFF4FBF87),
    agentPanelWarning = ZadMustardLight,
    agentPanelInfo = Color(0xFF7FB3E8),
    agentPanelBrand = Color(0xFF95D9B5),
    alertBannerContainer = Color(0xFFFDECEA),
    onAlertBanner = ZadTerracottaDark,
    aiAccent = Color(0xFF9333EA),
    aiAccentEnd = Color(0xFF6C63FF),
    onAiAccent = Color(0xFFFFFFFF),
    aiAccentSoft = Color(0xFFC084FC),
    aiNarrativeContainer = Color(0xFFF5F3FF),
    onAiNarrative = Color(0xFF4C1D95),
    scoreGood = Color(0xFF4D7C0F),
    onStatusPill = Color(0xFFA63F1B),
    // Exactly the values CategoryBreakdownCard used as ZadV3.green600 / info / warn /
    // danger / violet, in the same order — so the light rendering is unchanged.
    chartCategorical = listOf(
        Color(0xFF0F9B76),
        Color(0xFF2563EB),
        Color(0xFFB45309),
        Color(0xFFDC5B4B),
        Color(0xFF7C3AED),
    ),
    textOnCardSecondary = Color(0xFF374151),
)

/**
 * Dark counterparts. Semantics are preserved rather than the hex being inverted:
 * `success` stays a green that reads as success, `coralLight` stays the *container*
 * behind coral (so it goes dark, not pale), and `textTertiary` stays dimmer than
 * onSurfaceVariant so the three text weights keep their order.
 *
 * `primaryDark`/`primaryLight` keep their relative direction — darker and lighter than
 * `primary` — because both are used as gradient stops against it.
 */
val ZadExtendedColorsDark = ZadExtendedColors(
    success = Color(0xFF4FBF87),
    info = Color(0xFF7FB3E8),
    // كان 7F847A = 4.46:1 على الكروت — تحت الحد بسنة. 8E9388 = 5.43:1.
    textTertiary = Color(0xFF8E9388),
    primaryDark = Color(0xFF3E8F68),
    primaryLight = Color(0xFF95D9B5),
    secondaryDark = Color(0xFFC08A2C),
    secondaryLight = Color(0xFFF3C476),
    coral = ZadTerracottaLight,
    coralLight = Color(0xFF5A2A16),
    lilac = Color(0xFFB9A9D4),
    canvasTop = Color(0xFF10130F),
    canvasMid = Color(0xFF141712),
    canvasBottom = Color(0xFF191D17),
    surfaceContainerLow = Color(0xFF151813),
    surfaceContainer = ZadIosSurfaceDark,
    surfaceContainerHigh = Color(0xFF22261F),
    // Same role at dark luminance: a warm and a cool tint over the near-black canvas,
    // not the pastels — those wash out to a light screen, which is the bug being fixed.
    authWashWarm = Color(0xFF3A1E16),
    authWashCool = Color(0xFF14302A),
    agentPanelStart = Color(0xFF052E16),
    agentPanelEnd = Color(0xFF0A382C),
    agentPanelSuccess = Color(0xFF4FBF87),
    agentPanelWarning = ZadMustardLight,
    agentPanelInfo = Color(0xFF7FB3E8),
    agentPanelBrand = Color(0xFF95D9B5),
    alertBannerContainer = Color(0xFF3A1E16),
    onAlertBanner = ZadTerracottaOnDark,
    aiAccent = Color(0xFFC084FC),
    aiAccentEnd = Color(0xFF6C63FF),
    onAiAccent = Color(0xFF1E0B36),
    aiAccentSoft = Color(0xFFC084FC),
    aiNarrativeContainer = Color(0xFF261B3D),
    onAiNarrative = Color(0xFFDDD0FB),
    scoreGood = Color(0xFFA3E635),
    onStatusPill = Color(0xFFEA764B),
    // Same five hues, lifted to read on a dark ground. The light values are mid-tones
    // chosen against white; on near-black they lose separation and #B45309 in particular
    // goes muddy. Hue and order are preserved so a category keeps its colour identity
    // between themes — only luminance moves.
    chartCategorical = listOf(
        Color(0xFF34D399),
        Color(0xFF60A5FA),
        Color(0xFFE9A844),
        Color(0xFFF08A7A),
        Color(0xFFA78BFA),
    ),
    textOnCardSecondary = Color(0xFFB9BDB3),
)

val LocalZadExtendedColors = staticCompositionLocalOf { ZadExtendedColorsLight }

File: zad_flutter/lib/design/tokens/zad_extended_colors.dart
/// Kotlin's `ZadExtendedColors` — the colours Material's scheme has no slot
/// for (success, info, tertiary text, the canvas stops, the agent panel, the
/// AI identity…), with the same light and dark values.
///
/// Read it with `context.zadExt`, the way Kotlin reads
/// `LocalZadExtendedColors.current`. The per-field comments in
/// `ZadExtendedColors.kt` explain each value's contrast measurement; they are
/// not repeated here so there is one place to keep them true.
library;

// Each member carries its Kotlin name, and Kotlin's file documents it; a
// second copy of those comments here would only drift.
// ignore_for_file: public_member_api_docs

import 'package:flutter/material.dart';
import 'package:zad/design/tokens/zad_palette.dart';

/// Colours with no `ColorScheme` slot, per theme.
@immutable
class ZadExtendedColors extends ThemeExtension<ZadExtendedColors> {
  /// All fields, as in Kotlin's data class.
  const new({
    required this.success,
    required this.info,
    required this.textTertiary,
    required this.primaryDark,
    required this.primaryLight,
    required this.secondaryDark,
    required this.secondaryLight,
    required this.coral,
    required this.coralLight,
    required this.lilac,
    required this.canvasTop,
    required this.canvasMid,
    required this.canvasBottom,
    required this.surfaceContainerLow,
    required this.surfaceContainer,
    required this.surfaceContainerHigh,
    required this.authWashWarm,
    required this.authWashCool,
    required this.agentPanelStart,
    required this.agentPanelEnd,
    required this.agentPanelSuccess,
    required this.agentPanelWarning,
    required this.agentPanelInfo,
    required this.agentPanelBrand,
    required this.alertBannerContainer,
    required this.onAlertBanner,
    required this.aiAccent,
    required this.aiAccentEnd,
    required this.onAiAccent,
    required this.aiAccentSoft,
    required this.aiNarrativeContainer,
    required this.onAiNarrative,
    required this.scoreGood,
    required this.onStatusPill,
    required this.chartCategorical,
    required this.textOnCardSecondary,
  });

  final Color success;
  final Color info;
  final Color textTertiary;
  final Color primaryDark;
  final Color primaryLight;
  final Color secondaryDark;
  final Color secondaryLight;
  final Color coral;
  final Color coralLight;
  final Color lilac;
  final Color canvasTop;
  final Color canvasMid;
  final Color canvasBottom;
  final Color surfaceContainerLow;
  final Color surfaceContainer;
  final Color surfaceContainerHigh;
  final Color authWashWarm;
  final Color authWashCool;
  final Color agentPanelStart;
  final Color agentPanelEnd;
  final Color agentPanelSuccess;
  final Color agentPanelWarning;
  final Color agentPanelInfo;
  final Color agentPanelBrand;
  final Color alertBannerContainer;
  final Color onAlertBanner;
  final Color aiAccent;
  final Color aiAccentEnd;
  final Color onAiAccent;
  final Color aiAccentSoft;
  final Color aiNarrativeContainer;
  final Color onAiNarrative;
  final Color scoreGood;
  final Color onStatusPill;
  final List<Color> chartCategorical;
  final Color textOnCardSecondary;

  /// Kotlin's `ZadExtendedColorsLight`.
  static const ZadExtendedColors light = ZadExtendedColors(
    success: Color(0xFF238652),
    info: Color(0xFF2B6CB0),
    textTertiary: Color(0xFF6E7065),
    primaryDark: ZadPalette.forestEmeraldDark,
    primaryLight: ZadPalette.forestEmeraldLight,
    secondaryDark: ZadPalette.mustardDark,
    secondaryLight: ZadPalette.mustardLight,
    coral: ZadPalette.terracottaRust,
    coralLight: ZadPalette.terracottaContainer,
    lilac: Color(0xFF7C6F93),
    canvasTop: Color(0xFFF8F9FA),
    canvasMid: Color(0xFFF4F6F2),
    canvasBottom: Color(0xFFEDEFE9),
    surfaceContainerLow: Color(0xFFFBFBFA),
    surfaceContainer: ZadPalette.iosBackground,
    surfaceContainerHigh: ZadPalette.iosSurfaceVariant,
    authWashWarm: Color(0xFFFCD3C7),
    authWashCool: Color(0xFFBFE3D1),
    agentPanelStart: Color(0xFF052E16),
    agentPanelEnd: Color(0xFF0A382C),
    agentPanelSuccess: Color(0xFF4FBF87),
    agentPanelWarning: ZadPalette.mustardLight,
    agentPanelInfo: Color(0xFF7FB3E8),
    agentPanelBrand: Color(0xFF95D9B5),
    alertBannerContainer: Color(0xFFFDECEA),
    onAlertBanner: ZadPalette.terracottaDark,
    aiAccent: Color(0xFF9333EA),
    aiAccentEnd: Color(0xFF6C63FF),
    onAiAccent: Color(0xFFFFFFFF),
    aiAccentSoft: Color(0xFFC084FC),
    aiNarrativeContainer: Color(0xFFF5F3FF),
    onAiNarrative: Color(0xFF4C1D95),
    scoreGood: Color(0xFF4D7C0F),
    onStatusPill: Color(0xFFA63F1B),
    chartCategorical: <Color>[
      Color(0xFF0F9B76),
      Color(0xFF2563EB),
      Color(0xFFB45309),
      Color(0xFFDC5B4B),
      Color(0xFF7C3AED),
    ],
    textOnCardSecondary: Color(0xFF374151),
  );

  /// Kotlin's `ZadExtendedColorsDark`.
  static const ZadExtendedColors dark = ZadExtendedColors(
    success: Color(0xFF4FBF87),
    info: Color(0xFF7FB3E8),
    textTertiary: Color(0xFF8E9388),
    primaryDark: Color(0xFF3E8F68),
    primaryLight: Color(0xFF95D9B5),
    secondaryDark: Color(0xFFC08A2C),
    secondaryLight: Color(0xFFF3C476),
    coral: ZadPalette.terracottaLight,
    coralLight: Color(0xFF5A2A16),
    lilac: Color(0xFFB9A9D4),
    canvasTop: Color(0xFF10130F),
    canvasMid: Color(0xFF141712),
    canvasBottom: Color(0xFF191D17),
    surfaceContainerLow: Color(0xFF151813),
    surfaceContainer: ZadPalette.iosSurfaceDark,
    surfaceContainerHigh: Color(0xFF22261F),
    authWashWarm: Color(0xFF3A1E16),
    authWashCool: Color(0xFF14302A),
    agentPanelStart: Color(0xFF052E16),
    agentPanelEnd: Color(0xFF0A382C),
    agentPanelSuccess: Color(0xFF4FBF87),
    agentPanelWarning: ZadPalette.mustardLight,
    agentPanelInfo: Color(0xFF7FB3E8),
    agentPanelBrand: Color(0xFF95D9B5),
    alertBannerContainer: Color(0xFF3A1E16),
    onAlertBanner: ZadPalette.terracottaOnDark,
    aiAccent: Color(0xFFC084FC),
    aiAccentEnd: Color(0xFF6C63FF),
    onAiAccent: Color(0xFF1E0B36),
    aiAccentSoft: Color(0xFFC084FC),
    aiNarrativeContainer: Color(0xFF261B3D),
    onAiNarrative: Color(0xFFDDD0FB),
    scoreGood: Color(0xFFA3E635),
    onStatusPill: Color(0xFFEA764B),
    chartCategorical: <Color>[
      Color(0xFF34D399),
      Color(0xFF60A5FA),
      Color(0xFFE9A844),
      Color(0xFFF08A7A),
      Color(0xFFA78BFA),
    ],
    textOnCardSecondary: Color(0xFFB9BDB3),
  );

  @override
  ZadExtendedColors copyWith() => this;

  /// Theme switches are a swap, not a tween — Kotlin's
  /// `staticCompositionLocalOf` swaps the whole value too.
  @override
  ZadExtendedColors lerp(ZadExtendedColors? other, double t) =>
      t < 0.5 ? this : (other ?? this);
}

/// `context.zadExt.success`, as Kotlin writes `successColor`.
extension ZadExtendedColorsContext on BuildContext {
  /// The active theme's extended colours.
  ZadExtendedColors get zadExt =>
      Theme.of(this).extension<ZadExtendedColors>() ?? ZadExtendedColors.light;

  /// Kotlin's `isDarkSurface()` / `isSystemInDarkTheme()` as the theme sees it.
  bool get zadIsDark => Theme.of(this).brightness == Brightness.dark;
}
